agent/util/subtree_unix.go

```diff
@@ -0,0 +1,224 @@
+// +build darwin linux
+
+package util
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/grip"
+	"github.com/pkg/errors"
+)
+
+const (
+	cleanupCheckAttempts   = 10
+	cleanupCheckTimeoutMin = 100 * time.Millisecond
+	cleanupCheckTimeoutMax = 1 * time.Second
+	contextTimeout         = 10 * time.Second
+)
+
+// TrackProcess is a noop by default if we don't need to do any special
+// bookkeeping up-front.
+func TrackProcess(key string, pid int, logger grip.Journaler) {}
+
+// KillSpawnedProcs kills processes that descend from the agent and waits
+// for them to terminate
+func KillSpawnedProcs(ctx context.Context, key, workingDir string, logger grip.Journaler) error {
+	pidsToKill, err := getPIDsToKill(ctx, key, workingDir, logger)
+	if err != nil {
+		return errors.Wrap(err, "problem getting list of PIDs to kill")
+	}
+
+	for _, pid := range pidsToKill {
+		p := os.Process{Pid: pid}
+		err := p.Kill()
+		if err != nil {
+			logger.Errorf("Cleanup got error killing pid '%d': %v", pid, err)
+		} else {
+			logger.Infof("Cleanup killed pid '%d'", pid)
+		}
+	}
+
+	pidsStillRunning := waitForExit(ctx, pidsToKill)
+	for _, pid := range pidsStillRunning {
+		logger.Infof("Failed to clean up process '%d'", pid)
```

Comment:
[The internet](https://stackoverflow.com/a/767577) tells me this might happen if a process is in uninterruptible sleep. In this case the process may _eventually_ wake up, get the signal, and die, or it may never.
But do we really want to indiscriminately get rid of the host in this case? Particularly in the case of static hosts, it would be annoying if we quarantine a host only to find out the problem process exited a minute later.
I opened [EVG-15364](https://jira.mongodb.org/browse/EVG-15364) to think more about this.